test/unit/gettid_register.c

```diff
@@ -0,0 +1,121 @@
+/*
+ *  Copyright (c) 2018 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* Gettid Register Test: Register a custom gettid function pointer  */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <pthread.h>
+#include <shmem.h>
+#include <shmemx.h>
+
+#define T 8
+
+int dest;
+
+int me, npes;
+int errors = 0;
+
+
+uint64_t my_gettid(void) {
+    uint64_t tid_val = (uint64_t)pthread_self();
```

Comment:
You may want to add a sanity check in `main` to the effect:
```
if (sizeof(pthread_t) < sizeof(uint64_t)) {
  printf("Can't run this test, sorrynotsorry\n");
  shmem_finalize();
  return 0;
}
```

Reply:
Should that be `>`?  The concern is losing bits from the cast, right?

Also, it looks like a POSIX `pthread_t` could be a struct:
"IEEE Std 1003.1-2001/Cor 2-2004, item XBD/TC2/D6/26 is applied, adding pthread_t to the list of types that are not required to be arithmetic types, thus allowing pthread_t to be defined as a structure."

Maybe this cast will come back to bite us?

Reply:
It probably will (see comment below -- if it's a struct, you're likely to get non-unique TIDs by just looking at the first 8 bytes, which will break private STX assignment).  We do pass a `tid` as an argument to `thread_main`.  Why don't you move that into a global `_Thread_local int tid` variable and return that value from `my_gettid`?  `_Thread_local` is a C11-ism, so you should wrap this in an preprocessor check for C11 support and just return success if the compiler doesn't support C11.  Alternatively (and probably easier from a portability perspective), you can use `pthread_set/get_specific`.  Just create the key before you spawn threads to avoid needing the synchronize on key creation.

Reply:
👍 I'll look into the `pthread_setspecific` option.